### PR TITLE
http con man: set response flag on downstream protocol error

### LINKED ISSUE #8446: HTTP protocol error results in DC access log
The DC stream flag is set in the absence of any other stream flags/response codes: https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_impl.cc#L504

When a codec exception happens, we immediately reset the stream, setting neither a stream flag or a response code: https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_impl.cc#L263

As a result, protocol errors appear with the DC flag set in the access logs.

We've seen Envoy respond with a 400 for some protocol errors at least (in which case the response code should probably be set), but I'm guessing not all protocol errors result in a 400 so we should at least set a stream flag for protocol errors (PE?)

### PR DESCRIPTION
**Pull request by spenceral**
This adds a new response flag, DPE, that is set when the downstream
request has an HTTP protocol error.

Signed-off-by: Spencer Lewis <slewis@squareup.com>

Description: Protocol errors previously resulted in just the DC flag being set. Now, DPE will be set when there is a codec exception caused by the downstream request.
Risk Level: low
Testing: unit test
Docs Changes: update list of response flags
Release Notes: added release note
Fixes #8446

### CODE REVIEW

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -348,6 +349,9 @@ void ConnectionManagerImpl::resetAllStreams() {
     auto& stream = *streams_.front();
     stream.response_encoder_->getStream().removeCallbacks(stream);
     stream.onResetStream(StreamResetReason::ConnectionTermination, absl::string_view());
+    if (response_flag) {
+      stream.stream_info_.setResponseFlag(*response_flag);
```
Comment (junr03) [inline source/common/http/conn_manager_impl.cc]: nit: the style almost everywhere on the codebase is to use `has_value()` and `value()`

Comment (spenceral) [reply source/common/http/conn_manager_impl.cc]: Ah, that is nicer. Thanks

**Comment by spenceral** [comment]
Force pushed the last commit because I forgot to sign 🙃

**Comment by spenceral** [comment]
I think the failing test is a flake. When I locally run
```
bazel test //test/integration:http_subset_lb_integration_test --nocache_test_results --runs_per_test=100
```
I get
```
INFO: Build completed, 1 test FAILED, 101 total actions
//test/integration:http_subset_lb_integration_test                       FAILED in 1 out of 100 in 1.8s
```

**Comment by spenceral** [comment]
@junr03 is there a way to rerun the Azure tests?

**Comment by junr03** [comment]
@htuch this change lgtm, do you mind doing the last pass given it needs API approval?

**Comment by mattklein123** [comment]
Legit format failure. Please merge master and fix format.

/wait

**Pull request merged**